### PR TITLE
[b/383901867] Table API

### PR DESCRIPTION
**Pull request by InigoSJ**
Fixes b/383901867

Add Table API and examples

All the schema checks are happening when the serializer is created, so we don't need to do them in the Table Factory

### CODE REVIEW

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        BigtableDynamicTableSink other = (BigtableDynamicTableSink) obj;
+        return Objects.equals(resolvedSchema, other.resolvedSchema)
+                && Objects.equals(connectorOptions, other.connectorOptions);
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:91]: should we check that the rowkey field is the same here too?

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        DataType physicalSchema = resolvedSchema.toPhysicalRowDataType();
+
+        RowDataToRowMutationSerializer.Builder serializerBuilder =
+                RowDataToRowMutationSerializer.builder()
+                        .withSchema(physicalSchema)
+                        .withRowKeyField(this.rowKeyField);
+
+        if (connectorOptions.get(BigtableConnectorOptions.USE_NESTED_ROWS_MODE)) {
+            serializerBuilder.withNestedRowsMode();
+        } else {
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:110]: should we log here if we're defaulting to column family mode?

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+                        .setProjectId(connectorOptions.get(BigtableConnectorOptions.PROJECT))
+                        .setTable(connectorOptions.get(BigtableConnectorOptions.TABLE))
+                        .setInstanceId(connectorOptions.get(BigtableConnectorOptions.INSTANCE))
+                        .setSerializer(serializerBuilder.build())
+                        .build();
+
+        if (this.parallelism == null) {
+            return SinkV2Provider.of(bigtableSink);
+        }
+        return SinkV2Provider.of(bigtableSink, this.parallelism);
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:126]: i dont have a preference on if you explicitly reference this.variable vs just using variable for class variables but we should keep it consistent (ie, on like 137, we should use this.)

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/config/BigtableConnectorOptions.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    .booleanType()
+                    .defaultValue(false)
+                    .withDescription(
+                            "Specifies the use of nested rows as column families. Incompatible with `column-family`.");
+
+    public static final ConfigOption<Integer> SINK_PARALLELISM =
+            ConfigOptions.key("parallelism")
+                    .intType()
+                    .noDefaultValue()
+                    .withDescription("Sink Parallelism");
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/config/BigtableConnectorOptions.java:66]: nit: uncapitalized parallelism

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+                                    DataTypes.FIELD(TestingUtils.STRING_FIELD, DataTypes.STRING()),
+                                    DataTypes.FIELD(TestingUtils.INTEGER_FIELD, DataTypes.INT()))),
+                    Column.physical(
+                            TestingUtils.NESTED_COLUMN_FAMILY_2,
+                            DataTypes.ROW(
+                                    DataTypes.FIELD(
+                                            TestingUtils.STRING_FIELD_2, DataTypes.STRING()))));
+
+    @Test
+    public void testDynamicTableBigtableSink() throws IOException {
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:80]: we should also add tests to validate errors are throw when required options arent set correctly

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+                connectorOptions.get(BigtableConnectorOptions.INSTANCE), TestingUtils.INSTANCE);
+        assertEquals(connectorOptions.get(BigtableConnectorOptions.PROJECT), TestingUtils.PROJECT);
+
+        assertEquals(sink.rowKeyField, TestingUtils.ROW_KEY_FIELD);
+        assertEquals((Integer) sink.parallelism, (Integer) 2);
+        assertEquals(sink.resolvedSchema, schema);
+    }
+
+    @Test
+    public void testDynamicTableBigtableSinkNested() throws IOException {
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:131]: let's add a test to validate the behavior of USE_NESTED_ROWS_MODE but the schema is not a nested rows schema

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        options.put(BigtableConnectorOptions.USE_NESTED_ROWS_MODE.key(), "true");
+        options.put(BigtableConnectorOptions.SINK_PARALLELISM.key(), parallelism);
+
+        BigtableDynamicTableSink sink =
+                (BigtableDynamicTableSink) FactoryMocks.createTableSink(schema, options);
+
+        String summaryString =
+                String.format(
+                        "BigtableSink("
+                                + "parallelism=%s, "
```
Comment (clmccart) [inline connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:201]: doesnt need to be a test but how does string formatting work if parallelism is null?

File: connectors/bigtable/flink-examples-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/examples/WriteTableAPINested.java [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+        Schema schemaBigtable =
+                Schema.newBuilder()
+                        .column("my-key", DataTypes.STRING().notNull())
+                        .column(
+                                "family1",
+                                DataTypes.ROW(
+                                        DataTypes.FIELD("stringColumn", DataTypes.STRING()),
+                                        DataTypes.FIELD("intColumn", DataTypes.INT())))
+                        .column(
```
Comment (clmccart) [inline connectors/bigtable/flink-examples-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/examples/WriteTableAPINested.java:93]: this is kind of unrelated to this PR but what happens if a user is using nested rows mode and doesnt specify a column family for a column that is not nested? does that question make sense...

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+                        .getDataType()
+                        .equals(DataTypes.STRING().notNull()),
+                String.format(
+                        ErrorMessages.ROW_KEY_STRING_TYPE_TEMPLATE,
+                        resolvedSchema.getColumn(rowKeyIndex).get().getDataType()));
+
+        this.connectorOptions = connectorOptions;
+        this.resolvedSchema = resolvedSchema;
+        this.parallelism = connectorOptions.get(BigtableConnectorOptions.SINK_PARALLELISM);
+        this.rowKeyField = resolvedSchema.getColumn(rowKeyIndex).get().getName();
```
[the remark this answers is not in the document]
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:70]: It's a Flink native object that comes from Catalog.

The row key existing is checked before it:

```
        checkArgument(
                resolvedSchema.getPrimaryKeyIndexes().length == 1,
                String.format(
                        ErrorMessages.MULTIPLE_PRIMARY_KEYS_TEMPLATE,
                        resolvedSchema.getPrimaryKeyIndexes().length));
        int rowKeyIndex = resolvedSchema.getPrimaryKeyIndexes()[0];
        checkArgument(
                resolvedSchema
                        .getColumn(rowKeyIndex)
                        .get()
                        .getDataType()
                        .equals(DataTypes.STRING().notNull()),
                String.format(
                        ErrorMessages.ROW_KEY_STRING_TYPE_TEMPLATE,
                        resolvedSchema.getColumn(rowKeyIndex).get().getDataType()));
```

There are tests already for all specific cases for the Sink (testRowKeyValidation). The only case that is not covered is if the Primary Key is not in the schema, but this error is handled by the factory and it's Flink Native. 

I could add it but I don't think it covers anything that is not covered already

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    public int hashCode() {
+        return Objects.hash(this.resolvedSchema, this.connectorOptions);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
```
[the remark this answers is not in the document]
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:83]: I think the current order is the fastest:

this == obj => This is checking if the two objects are the same instance in memory. This should be the fastest

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 1 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:91]: it's in the schema

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 2 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:110]: Users should know they are using column family as it needs to be set.

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 3 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:126]: Thanks!

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java (hunk 6 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:131]: I should ahve add this in the PR, sorry:

All the schema checks are happening when the serializer is created, so we don't need to do them in the Table Factory

File: connectors/bigtable/flink-examples-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/examples/WriteTableAPINested.java (hunk 8 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-examples-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/examples/WriteTableAPINested.java:93]: All of that is tested, bothn in Generic and RowData. For Table API, it uses RowData tests as it uses it as serializer:

https://github.com/google/flink-connector-gcp/blob/main/connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/serializers/RowDataTest.java#L147

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java (hunk 7 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:201]: Sure, I thought it was too much hahaha. Added

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.resolvedSchema, this.connectorOptions);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
```
[the remark this answers is not in the document]
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:80]: Two objects being the same object (same address)

File: connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java (hunk 5 above)
Comment (InigoSJ) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/test/java/com/google/flink/connector/gcp/bigtable/table/BigtableTableTest.java:80]: I didn't know this was possible since it was a Flink native error (helper.validate()), I was about to give up but managed to do it. Thanks!

File: connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (clmccart) [reply connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java:70]: that's right you already told me this. sorry :)
